Test Failure: DataStoreTestServlet.testDefaultDataSourceByJNDIName

Test Failure: test.jakarta.data.datastore.DataStoreTest.DataStoreTestServlet.testDefaultDataSourceByJNDIName

```
DataStoreTestServlet.testDefaultDataSourceByJNDIName:junit.framework.AssertionFailedError: 2024-02-29-19:24:54:332 ERROR: Caught exception attempting to call test method testDefaultDataSourceByJNDIName on servlet test.jakarta.data.datastore.web.DataStoreTestServlet
jakarta.data.exceptions.MappingException: java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Not found: application[DataStoreTestApp]/module[DataStoreTestWeb.war]/databaseStore[java:comp/DefaultDataSource]
    at io.openliberty.data.internal.persistence.RepositoryImpl.failure(RepositoryImpl.java:549)
    at io.openliberty.data.internal.persistence.RepositoryImpl.invoke(RepositoryImpl.java:2749)
    at jdk.proxy11/jdk.proxy11.$Proxy63.existsByIdAndValue(Unknown Source)
    at test.jakarta.data.datastore.web.DataStoreTestServlet.testDefaultDataSourceByJNDIName(DataStoreTestServlet.java:63)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.app.FATServlet.doGet(FATServlet.java:73)
    at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:527)
    at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1260)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:748)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:445)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1361)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1077)
    at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:77)
    at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:87)
    at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:969)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:293)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1260)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:476)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:435)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:569)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:503)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:363)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:330)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:169)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:77)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:516)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:586)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:970)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1059)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:280)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:858)
Caused by: java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Not found: application[DataStoreTestApp]/module[DataStoreTestWeb.war]/databaseStore[java:comp/DefaultDataSource]
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
    at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
    at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1294)
    at java.base/java.util.concurrent.CompletableFuture.biApplyStage(CompletableFuture.java:1329)
    at java.base/java.util.concurrent.CompletableFuture.thenCombine(CompletableFuture.java:2210)
    at io.openliberty.data.internal.persistence.RepositoryImpl.<init>(RepositoryImpl.java:207)
    at io.openliberty.data.internal.persistence.cdi.RepositoryProducer$Factory.getHandler(RepositoryProducer.java:96)
    at io.openliberty.data.internal.persistence.cdi.RepositoryProducer.produce(RepositoryProducer.java:168)
    at org.jboss.weld.bean.SyntheticProducerBean.create(SyntheticProducerBean.java:44)
    at org.jboss.weld.contexts.unbound.DependentContextImpl.get(DependentContextImpl.java:64)
    at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
    at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
    at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:680)
    at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:780)
    at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
    at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:345)
    at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:356)
    at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:69)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl$3.run(WebSphereInjectionServicesImpl.java:325)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl$3.run(WebSphereInjectionServicesImpl.java:322)
    at java.base/java.security.AccessController.doPrivileged(AccessController.java:692)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl.aroundInject(WebSphereInjectionServicesImpl.java:322)
    at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:46)
    at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:71)
    at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.cdiInjection(CDIManagedObject.java:204)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.inject(CDIManagedObject.java:177)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.inject(CDIManagedObject.java:155)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.inject(WebApp.java:1278)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.injectAndPostConstruct(WebApp.java:1419)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.injectAndPostConstruct(WebApp.java:1407)
    at com.ibm.ws.webcontainer.osgi.servlet.ServletWrapper.createTarget(ServletWrapper.java:65)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper$1.run(ServletWrapper.java:1545)
    at java.base/java.security.AccessController.doPrivileged(AccessController.java:748)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.loadServlet(ServletWrapper.java:1513)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:592)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:445)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1361)
    at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5095)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:328)
    at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1038)
Caused by: java.lang.IllegalArgumentException: Not found: application[DataStoreTestApp]/module[DataStoreTestWeb.war]/databaseStore[java:comp/DefaultDataSource]
    at io.openliberty.data.internal.persistence.service.DBStoreEMBuilder.initialize(DBStoreEMBuilder.java:499)
    at io.openliberty.data.internal.persistence.EntityManagerBuilder.run(EntityManagerBuilder.java:177)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:280)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)

    at componenttest.topology.utils.FATServletClient.assertTestResponse(FATServletClient.java:103)
    at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:88)
    at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:44)
    at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:203)
    at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:363)
    at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:177)
```

This is in intermittent error where the service reference sometimes isn't available yet.

--- a/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/service/DBStoreEMBuilder.java
+++ b/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/service/DBStoreEMBuilder.java
@@ -46,6 +46,7 @@ import java.util.TreeMap;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
 
 import org.objectweb.asm.ClassWriter;
 import org.objectweb.asm.FieldVisitor;
@@ -86,6 +87,7 @@ import jakarta.persistence.EntityManager;
  */
 public class DBStoreEMBuilder extends EntityManagerBuilder {
     private static final String EOLN = String.format("%n");
+    private static final long MAX_WAIT_FOR_SERVICE_NS = TimeUnit.SECONDS.toNanos(60);
     private static final TraceComponent tc = Tr.register(DBStoreEMBuilder.class);
 
     private final ClassDefiner classDefiner = new ClassDefiner();
@@ -493,12 +495,25 @@ public class DBStoreEMBuilder extends EntityManagerBuilder {
         final boolean trace = TraceComponent.isAnyTracingEnabled();
 
         BundleContext bc = FrameworkUtil.getBundle(DatabaseStore.class).getBundleContext();
-        Collection<ServiceReference<DatabaseStore>> refs = bc.getServiceReferences(DatabaseStore.class,
-                                                                                   FilterUtils.createPropertyFilter("id", databaseStoreId));
-        if (refs.isEmpty())
-            throw new RuntimeException("Not found: " + databaseStoreId);
 
-        ServiceReference<DatabaseStore> ref = refs.iterator().next();
+        ServiceReference<DatabaseStore> ref = null;
+        for (long start = System.nanoTime(), poll_ms = 125L; ref == null; poll_ms = poll_ms < 1000L ? poll_ms * 2 : 1000L) {
+            Collection<ServiceReference<DatabaseStore>> refs = bc.getServiceReferences(DatabaseStore.class,
+                                                                                       FilterUtils.createPropertyFilter("id", databaseStoreId));
+            if (refs.isEmpty()) {
+                if (System.nanoTime() - start < MAX_WAIT_FOR_SERVICE_NS) {
+                    if (trace && tc.isDebugEnabled())
+                        Tr.debug(this, tc, "Wait " + poll_ms + " ms for service reference to become available...");
+                    TimeUnit.MILLISECONDS.sleep(poll_ms);
+                } else {
+                    throw new IllegalStateException("The " + databaseStoreId + " service component did not become available within " +
+                                                    TimeUnit.NANOSECONDS.toSeconds(MAX_WAIT_FOR_SERVICE_NS) + " seconds.");
+                }
+            } else {
+                ref = refs.iterator().next();
+            }
+        }
+
         String tablePrefix = (String) ref.getProperty("tablePrefix");
 
         if (trace && tc.isDebugEnabled())

